openhtf/io/output/print_console.py

```diff
@@ -0,0 +1,81 @@
+  """Return an output callback that writes JSON Test Records.
+
+  An example filename_pattern might be:
+    '/data/test_records/%(dut_id)s.%(start_time_millis)s'
+
+  To use this output mechanism:
+    test = openhtf.Test(PhaseOne, PhaseTwo)
+    test.AddOutputCallback(openhtf.OutputToJson(
+        '/data/test_records/%(dut_id)s.%(start_time_millis)s'))
+
+  Args:
+    filename_pattern: A format string specifying the filename to write to,
+      will be formatted with the Test Record as a dictionary.
+    inline_attachments: Whether attachments should be included inline in the
+      output.  Set to False if you expect to have large binary attachments.
+  """
+
+  def __init__(self, filename_pattern=None, inline_attachments=True, **kwargs):
+    super(PrintConsole, self).__init__(**kwargs)
+    self.FAIL =  '\033[91m'
+    self.PASS =  '\033[92m'
+    self.ERROR = '\033[93m'
+    self.ENDC =  '\033[0m'
+    self.BOLD =  '\033[1m'
+
+  # pylint: disable=invalid-name
+  def __call__(self, test_record):
+    as_dict = util.convert_to_dict(test_record)
+    result = as_dict['outcome']
+    color = None
+    if result == 'PASS':
+      color = self.PASS
+    elif result == 'FAIL':
+      color = self.FAIL
+    else:
+      color = self.ERROR
+
+    print '\n'+color+self.BOLD+os.path.basename(sys.argv[0])+": "+result+self.ENDC
+      
+    if result == 'FAIL' or result == 'ERROR':
+      phases=as_dict['phases']
+      for phase in phases:
+        phase_name = phase['name']
+        phase_time = str((float(phase['end_time_millis']) - \
+                               float(phase['start_time_millis']))/1000)
+        if result == 'FAIL':
+          measured = phase['measured_values']
+          measurements = phase['measurements']
+          for measurement in measurements:
+            res = measurements[measurement]
+            if res['outcome'] == 'FAIL':
+              if len(phase_name) > 0:
+                print '\nfailed phase: ' + phase_name + ' [time taken(s): ' + \
+                                                      phase_time+']'+self.ENDC
+                phase_name = ''
+
+              failed_item = res['name']
+              print ' '*4 + 'failed_item: ' + failed_item + self.ENDC
+              print ' '*8 + 'measured_value: '+ str(measured[failed_item]) + self.ENDC 
+              if 'validators' in res:
+                print ' '*8 + 'validator: ' + res['validators'][0]
+              else:
+                print ' '*8 + 'validator: None'
+        else:
+          phase_result = phase['result']
+          if phase_result['raised_exception'] == 'True':
+            print 'raised_exception phase: ' + phase_name +' [time taken(s): '\
+                              + phase_time + ']' + self.ENDC
+            exp_type = as_dict['outcome_details'][0]['code']
+            print ' '*4 + 'exception type : ' + exp_type + self.ENDC
+            print ' '*4 + 'exception text: ' + phase_result['phase_result'] + self.ENDC
```

Comment:
rather than have the print statements inline, it would be better to start with a list of `output_lines = []`, and then `output_lines.append(...)` each place you would print, and then at the end do `sys.stdout.write(''.join(output_lines))`